### ISSUE TITLE
Unhandled rejection when using callbacks (i.e. with `.watch(spec)`)

### ISSUE THREAD

**Issue by shicks** [metalsmith/metalsmith#393]
**Description**
When using the callback API (which is required with `Metalsmith.prototype.watch(spec)`), if the initial build results in an error, the error goes through as an unhandled rejection, rather than calling the callback with the error.

Subsequent builds eventually start to have the correct behavior: specifically, the second build will call the callback and _then_ drop an unhandled rejection (if both the first two failed), while the third and later builds only call the callback, as expected.  Unfortunately, without registering a global handler for the unhandled rejection (which users should _not_ have to do), the builds never get that far in the first place.

**Environment**

- OS: darwin
- Node|npm version: 18.12.1 | 8.19.2
- Metalsmith version: 2.6.1

**Reproducing the bug**

To reproduce, save the following file as "metalsmith.js".  You can tweak whether it will crash or not by adjusting the comment at the top, and run it with `node metalsmith.js [--watch|--promise] [--handle]` to see the various behaviors.  Note in particular that when using the Promise API, everything works great, but if a callback is passed, the first failure results in an unhandled rejection, and `watch()`ing is incompatible with promises, so this triggers the issue.

```javascript
// crash: true

const Metalsmith = require('metalsmith');

const watch = process.argv.includes('--watch');
const promise = process.argv.includes('--promise');
const handle = process.argv.includes('--handle');

if (handle) {
  process.on('unhandledRejection', (error, p) => {
    // Will print "unhandledRejection err is not defined"
    console.error(`Uncaught: ${error.message}`);
  });
}

function plugin(files, metalsmith, done) {
  for (const [name, file] of Object.entries(files)) {
    if (/\/\/ crash: true/.test(file.contents)) throw new Error('crash');
  }
  done();
}

const metalsmith = Metalsmith(__dirname)
    .source('.')
    .destination('build')
    .use(plugin);
if (watch) {
  metalsmith.watch('metalsmith.js');
}
if (promise) {
  (async () => {
    try {
      await metalsmith.build();
      console.log('Build succeeded');
    } catch (err) {
      console.log(`Build failed: ${err.message}`);
    }
  })();
} else {
  metalsmith.build((err) => {
    if (err) {
      console.log(`Build failed: ${err.message}`);
    } else {
      console.log('Build succeeded');
    }
  });
}
```

**Comment by shicks** [comment]
I've traced down two different places where unhandled rejections are happening.  I believe #394 fixes the issue.  I tried a few different places to handle the rejection (such as in the `return new Promise(...)` at the end of `watcher.js`), but many of the other options resulted in redundant calls to `callback`.  I was unable to get any duplicate calls with this PR, but I don't fully understand the watch workflow, so it's possible I missed a case.
